apps/settings/js/modules/bluetooth/bluetooth_context.js

```diff
@@ -0,0 +1,852 @@
+      });
+    },
+
+    /**
+     * A function to receive BluetoothDiscoveryHandle. And set a function to
+     * handle 'ondevicefound' event.
+     *
+     * @access private
+     * @memberOf BluetoothContext
+     * @param {Object BluetoothDiscoveryHandle} handle
+     */
+    _setDiscoveryHandler: function btc__setDiscoveryHandler(handle) {
+      this.debug('_setDiscoveryHandler(): handle = ' + handle);
+      // make the code base easy to do unit test
+      this._discoveryHandler = handle;
+      this._discoveryHandler.ondevicefound = this._onDeviceFound.bind(this);
+    },
+
+    /**
+     * Handle 'ondevicefound' event to access remote device in list with
+     * observable object.
+     *
+     * @access private
+     * @memberOf BluetoothContext
+     * @param {Object} evt
+     */
+    _onDeviceFound: function btc__onDeviceFound(evt) {
+      // save device
+      this._saveDevice(evt.device);
+    },
+
+    /**
+     * To distinguish between paired and unpaired.
+     * Then, save/update the device in list corresponding to the state of
+     * property paired.
+     *
+     * @access private
+     * @memberOf BluetoothContext
+     * @param {Object Observable} device
+     */
+    _saveDevice: function btc__saveDevice(device) {
+      // distinguish the found device is paired or not
+      var operatingDevices =
+        (device.paired) ? this.getPairedDevices() : this.getRemoteDevices();
+
+      // check the device is existed or not in remote/paired devices array
+      var index = operatingDevices.array.findIndex(
+                    this._findDeviceByAddress.bind(this, device.address));
+
+      // If the device is not existed yet, create observable oject
+      // for saving this device.
+      if (index === -1) {
+        // Get Bluetooth Device module.
+        this._getBluetoothDevice().then(function(BtDevice) {
+          // create observable BtDevice
+          var observableBtDevice = BtDevice(device);
+          // push device in devices list with observable object
+          operatingDevices.push(observableBtDevice);
+        });
+      } else {
+        // The device is existed, no need to do any thing here.
+        // set device in devices list
+        // operatingDevices.set(index, device);
+      }
+    },
+
+    /**
+     * Lazy loading BluetoothDevice module.
+     *
+     * @access private
+     * @memberOf BluetoothContext
+     * @return {Promise}
+     */
+    _getBluetoothDevice: function btc__getBluetoothDevice() {
+      return new Promise(function(resolve) {
+        require(['modules/bluetooth/bluetooth_device'], function(BtDevice) {
+          resolve(BtDevice);
+        });
+      });
+    },
+
+    /**
+     * Given address to find out device element from array.
+     *
+     * @access private
+     * @memberOf BluetoothContext
+     * @param {String} address
+     * @return {Boolean}
+     */
+    _findDeviceByAddress:
+    function btc__findDeviceByAddress(address, element) {
```

Comment:
nit: The name `element` is misleading as it should represent a bluetooth device.